Shorter sign up - remove Send Link by email screen

ParaskP7 opened the pull request:
Fixes #13971
Relates [WordPressAuthenticator-iOS#574](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/574)

This PR removes the below intermediate `Sign Up` screen:

<img src="https://user-images.githubusercontent.com/9729923/107496782-6ba45700-6b9a-11eb-8afd-9ce685baaecd.png" width="250" height="500">

Instead, an email will be now sent immediately after the user enters their email:

https://user-images.githubusercontent.com/9729923/107495717-203d7900-6b99-11eb-94bb-6393162595e3.mp4

ℹ️ An additional paragraph was added on the confirmation screen stating the: `Didn't mean to create a new account? Go back to re-enter your email address.`

⚠️ Since the `gotUnregisteredEmail(...)` functionality is shared between the `WordPress` and `WooCommerce` apps, this was not removed altogether. Instead the implementing client just end up doing nothing as an implementation on the `WordPress` side.

To test:
- Verify that the login and signup flows are working as expected, with no issue and with the above flow improvement.
- Check tracking is as expected.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ParaskP7 committed:
Feature: Replace unregistered email with show signup magic link step

This change effectively removes the intermediate 'Send link by email'
screen.

This screen is actually a dual purpose screen because it is also used
for with social login (Google). As such, the screen is not removed per
se, but actually becomes now a single purpose screen to be used only for
the purpose of logging in with Google.

ParaskP7 committed:
Feature: Add extra message to the email confirmation screen

This extra message will let the user know they can go back if they
didn't intend to sign up.

ParaskP7 committed:
Feature: Add extra message to the email confirmation screen

This extra message will let the user know they can go back if they
didn't intend to sign up.

ParaskP7 committed:
Refactor: Remove unused got unregistered email login interface method

ParaskP7 committed:
Refactor: Remove unused got unregistered email login impl. method

ParaskP7 committed:
Refactor: Simplify signup confirmation screen for use with social login

1) This screen no longer needs the extra 'newInstance(email)' method,
which was only used within the deleted 'gotUnregisteredEmail(...)'
method.
2) As such the 'mIsSocialSignup' and 'ARG_IS_SOCIAL_SIGNUP' logic is no
longer needed and deleted as well.
3) The tracking of 'trackEmailSignupConfirmationViewed()' is also no
needed based on the above changes.
4) Finally, the XML file can be updated to account only for the social
logic (Google).
5) And as part of that last change, the
'signup_confirmation_magic_link_message' string becomes obsolete as
well.

ParaskP7 committed:
Refactor: Remove unused track email signup confirmation viewed method

ParaskP7 committed:
Refactor: Remove unused signup confirmation magic link message string

ParaskP7 committed:
Refactor: Add back unused got unregistered email login interface method

This 'gotUnregisteredEmail(...)' method is indeed now unused for
the 'WordPress' app, but not for the 'WooCommerce' app. As such, this is
added back as an interface method to be shared with both apps. The only
difference is that for the 'WordPress' app it will not be doing anything
as of now.

ParaskP7 committed:
Refactor: Add back unused got unregistered email login impl. method

This 'gotUnregisteredEmail(...)' method is added back but it will not be
doing anything as of now.

ParaskP7 commented:
👋 @anitaa1990 !

Once more, I want to ping you on that change so that you are aware of it as well. After checking with `Woo` I understand that you are implementing this differently on your side, so I don't think you need to be doing much, but I figured it would be good if you are also aware of this.

ParaskP7 commented:
👋 @oguzkocer !

When this PR ends up being merged in `develop`, and since I am having issues pushing changes to the [WordPress-Login-Flow-Android](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android) , should I be doing it manually once more, creating a manual PR to that repo in order to apply the `Login` related changes?

PS: Previous [manual PR](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/50) on [WordPress-Login-Flow-Android](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android).

oguzkocer commented:
@ParaskP7 Yes, please go ahead with the manual update.

mattmiklic reviewed (approved):
Thanks for the ping; everything is looking as it should to me here.

renanferrari commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginEmailFragment.java:
```diff
@@ -717,7 +717,7 @@ public void onAuthOptionsFetched(OnAuthOptionsFetched event) {
                     if (mIsSignupFromLoginEnabled || isWooWPcomLoginFlow) {
                         if (mLoginListener != null) {
-                            mLoginListener.gotUnregisteredEmail(email);
+                            mLoginListener.showSignupMagicLink(email);
```
This changes the behavior for the Woo app as well, and they might not want to show the same screen we do at this point (in fact, [they don't](https://github.com/woocommerce/woocommerce-android/blob/5e8ca2f636f92260d29c5467256df1c4a3def812/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/LoginActivity.kt#L634-L641)).

Instead, I'd suggest to keep the call to `LoginListener::gotUnregisteredEmail` here and then use it to instantiate the `SignupMagicLinkFragment`.

ParaskP7 replied:
👋 @renanferrari !

Good catch! I actually did see `Woo` not using the `SignupConfirmationFragment` screen, but I missed the fact that this `gotUnregisteredEmail(...)` is being utilised on both apps. I'll do as you suggested, I'll revert this and make sure that `gotUnregisteredEmail(...)` on the `WordPress` side redirects to the `showSignupMagicLink(...)` instead.

PS: It is a bit tricky having to deal with these flows and the more `Woo` and `WordPress` diverge on their flows, compared to the original needs of this library, I suggest that at some point we revisit this library to make it fit the current needs.

ParaskP7 replied:
This is now done here: beff291750260a41bda15ae3476ed4d1fb120f01

renanferrari replied:
>It is a bit tricky having to deal with these flows and the more `Woo` and `WordPress` diverge on their flows, compared to the original needs of this library, I suggest that at some point we revisit this library to make it fit the current needs.

Totally agree! I believe there's a lot of room for improvement on how we share certain things with this library.

renanferrari commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginAnalyticsListener.kt:
```diff
@@ -65,7 +65,6 @@ interface LoginAnalyticsListener {
     fun trackPickEmailFromHint()
     fun trackShowEmailHints()
     fun emailFormScreenResumed()
-    fun trackEmailSignupConfirmationViewed()
```
This is a breaking change to the Woo app. It's not a big deal though, as long as you make sure to highlight it when pushing these changes to the library.

ParaskP7 replied:
👋 @renanferrari !

Oh you're right, again nice catch! `Woo` implements this method but as a `TODO("Not yet implemented")`, which means that when they pull the changes from this library the compiler will complain. I'll revert that change on the library level and have the `WordPress` app implement it with a `\\ Do nothing` comment. Wdyt?

ParaskP7 replied:
This is now done here: bbf246401b97edb50bfdd794b82d6fa4425cdad6

renanferrari replied:
In this case I'm wondering if it's not better to actually remove this, since we are effectively removing the "signup confirmation" screen for email signups. And as none of the apps are going to use it, it seems right. What do you think?

ParaskP7 replied:
👋 @renanferrari !

Do you mean `if it's better to actually remove this`, without the `not`, I just got a bit confused? 😅 

If this is indeed what you meant, I was thinking along the same lines, but decided that it probably better to not have that removed since on the `Woo` side it is marked as `TODO("Not yet implemented")`. We now know that on the `WordPress` side this is a `\\ Do nothing.`, but I am not sure about `Woo`.

Also `Woo` has an intermediate `LoginNoWPcomAccountFoundFragment` screen as well, which made me think that I shouldn't remove this logic from `Woo` just yet (it might be of used as some point, or planned).

<img src="https://user-images.githubusercontent.com/9729923/107741305-936bfa00-6d15-11eb-9b42-5831b599d1e2.png" width="250" height="500">

However, if you really think that it is indeed better if we just remove this, and have `Woo` update their implementation by removing this as well, that is when they pull the latest changes, then I will confidently do so and I would agree. It is just that I choose the less risky, frictionless path on this instance.

ParaskP7 replied:
👋 @renanferrari !

As per our discussion, I removed this tracking event, you can check this commit: 5252eda1c835abb4d51eafed864db6dfe6fb20df

ParaskP7 replied:
👋 @anitaa1990 !

Could you also please verify that this is okay on your end?

Many thanks in advance! 🙏

renanferrari reviewed (commented):
Hey @ParaskP7, thanks for the changes!

I've tested the app and it seems to be working as expected. That said, it looks like this introduces some changes that could impact the Woo app, so unless you’ve already coordinated this with them, you should probably make some considerations, which I’ve detailed below.

ParaskP7 committed:
Refactor: Add back the got unregistered email call on login email frag.

This is done purely because otherwise the behavior for the 'Woo' app
will change, which will end up breaking the 'Woo' flows.

As such, this commit reverts that and instead updates the
'LoginActivity', which exists only on the 'WordPress' side and makes
sure that the implementing 'gotUnregisteredEmail(...)' method calls the
'showSignupMagicLink(...)' instead, thus updating only the 'WordPress'
login flow with the intended change.

ParaskP7 committed:
Refactor: Add back the track email signup confirmation viewed method

This is done purely because otherwise this will break the 'Woo' app,
which also using this interface method, even if that is currently just a
'TODO("Not yet implemented")'.

As such, this commit reverts that and instead updates the implementing
method on the 'WordPress' side with a '\\ Do nothing.' comment.

ParaskP7 commented:
👋 @renanferrari !

So many thanks for the excellent review, especially when it comes to how this will affect `Woo`! 🙏 

> I've tested the app and it seems to be working as expected. That said, it looks like this introduces some changes that could impact the Woo app, so unless you’ve already coordinated this with them, you should probably make some considerations, which I’ve detailed below.

You're right! I have made some updates based on your comments, you could please take another look. 🙏

ParaskP7 committed:
Refactor: Remove unused track email signup confirmation viewed method

After some more thought, it has been decided to remove this and have
'Woo' update their app when pulling this login library change.

renanferrari reviewed (approved):
Thanks for the changes. Feel free to merge this once @anitaa1990 verifies everything is ok (I believe it should be).

Great job on this! LGTM :shipit:

anitaa1990 commented:
Hey @ParaskP7 👋 Thank you for the ping! 

Is this screen removed only during the sign up process? If that is the case, we currently don't support sign ups in the Woo app so you should be good to go 👍

ParaskP7 commented:
👋 @anitaa1990 !

> Is this screen removed only during the sign up process?

Yes, you are right Anitaa, this screen is used during the sign up process.

> If that is the case, we currently don't support sign ups in the Woo app so you should be good to go 👍

Perfect, thanks for the confirmation Anitaa, will now merge this PR! 🙏

ParaskP7 commented:
PS: A [manual PR](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/51) was created (cherry-peeking the commit from this PR) within the [WordPress-Login-Flow-Android](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android) repo to make sure that this change is replicated within that repo as well.
